lib/iris/exceptions.py

```diff
@@ -180,3 +183,19 @@ class CannotAddError(ValueError):
     """Raised when an object (e.g. coord) cannot be added to a :class:`~iris.cube.Cube`."""
 
     pass
+
+
+@lru_cache(None)
+def warn_once(msg, type, stacklevel, frame):
+    """Raise a warning only if a similar one has not been raised before."""
+    warnings.warn(msg, type, stacklevel=stacklevel)
+
+
+def warn_once_at_level(msg, type, stacklevel):
+    """
+    Raise a warning only if a similar one hasn't been raised from the same line
+    for a given stack level.
+    """
+    stacklevel += 1
+    frame = traceback.format_stack()[-stacklevel]
+    warn_once(msg, type, stacklevel + 1, frame)
```

Comment:
@stephenworsley Would you consider namespacing this in `iris.warnings` instead?

Seems a tad odd for `iris.exceptions` to raise a warning ... initially I'd not expect that behaviour to live there.

Reply:
I'm not against this; just wanted to point out there's a lot of other stuff we should move as well, if we do this!

Reply:
Seems like we should consider doing that then if we're in this space, otherwise it won't happen ... typically, that's the case.

Reply:
We did this! #5760